### PR TITLE
feat(new transform): Add `metric_to_log` transform

### LINKED ISSUE #1199: New `metric_to_log` transform
I can see use cases where users would want to ingest raw metric data into an events/log database. Providing a simple way for users to convert metric to log data would allow them to use these storages. For example, sending metrics data to S3 for archival and then using Athena to query it.

### PR DESCRIPTION
**Pull request by drunkirishcoder**
Closes #1199 

Signed-off-by: drunkirishcoder <daniel.jin@gmail.com>

### CODE REVIEW

**Review by bruceg — CHANGES_REQUESTED** [review]
The code here looks pretty good. The PR needs to have some documentation though (see `.meta/transforms`), and we should probably have at least a little discussion about the data model for the generated log. This seems reasonable but I'd like confirmation. Ping @lukesteensen

**Comment by drunkirishcoder** [comment]
Added the documentation for the transform. Open to discussion on the log data model.

**Review by lukesteensen — COMMENTED** [review]
My initial thought is probably that we should match the existing JSON representation of our metrics data, which could also make for a convenient implementation. We'd want to make sure it's performant enough, but a simple `metric event -> json -> log event` could be a good starting point here?

**Comment by drunkirishcoder** [comment]
Yeah I basically replicated the JSON representation of the metrics data. But I didn't do the to/from JSON conversion thinking that manually mapping the fields from metric to log event would perform better. maybe?

**Comment by lukesteensen** [comment]
It likely does perform a bit better, but I'm curious how big of a difference it makes. If we can make the code significantly simpler and more robust to a changing data model, that's probably worth a few percent different in performance (at least for the time being).

**Comment by drunkirishcoder** [comment]
Yeah makes sense, less code is always preferred for sure.

I guess my only other concern here is the possibility of the internal metrics representation changing. So the reason we wanted this feature is we are planning to use Vector to scrape prometheus endpoints, then convert to logs, and send them to Elasticsearch. We will build dashboards with the data. I guess the worry here is less to do with JSON serialization perf, more about not having a public contract of the output JSON format, and serialization could silently break that when the internal structure changes.

**Comment by binarylogic** [comment]
Thanks @drunkirishcoder, I've opened https://github.com/timberio/vector/issues/3779 to get consensus on this because I agree, we should think about how to do this properly.

**Comment by binarylogic** [comment]
Noting that we've decided to move forward with the conversion as-is here. The relevant discussion is in https://github.com/timberio/vector/issues/3779#issuecomment-689836306 if you have any comments to adds. Otherwise, this is unblocked and free to move forward.

**Comment by drunkirishcoder** [comment]
@binarylogic thanks, I will catch up on the convo and work on a serialize/deserialize version of the transform.

**Comment by binarylogic** [comment]
@drunkirishcoder we just merged https://github.com/timberio/vector/pull/4087.

To clarify, this contribution and any future contributions fall under the DCO.

**Comment by drunkirishcoder** [comment]
@bruceg @lukesteensen reimplemented the solution based on the feedback. It is now serializing from the `Metric` event.

File: src/transforms/metric_to_log.rs [hunk 1]
```diff
@@ -0,0 +1,346 @@
...
+        serde_json::to_vec(&metric)
+            .map_err(|error| emit!(MetricToLogFailedSerialize { error }))
+            .ok()
+            .and_then(|buffer| {
+                serde_json::from_slice::<Value>(buffer.as_slice())
+                    .map_err(|error| emit!(MetricToLogFailedDeserialize { error }))
+                    .ok()
+            })
+            .and_then(|value| {
+                if let Value::Object(object) = value {
```
Comment (bruceg) [inline src/transforms/metric_to_log.rs:75]:
```suggestion
                match value {
                    Value::Object(object) => {
```

**Review by bruceg — APPROVED** [review]
I too am concerned about the conversion of Metric -> Vec<u8> -> LogEvent, but I understand the logic for doing it this way.

File: src/transforms/metric_to_log.rs [hunk 2]
```diff
@@ -0,0 +1,346 @@
...
+        }
+    }
+}
+
+impl Transform for MetricToLog {
+    fn transform(&mut self, event: Event) -> Option<Event> {
+        let metric = event.into_metric();
+        emit!(MetricToLogEventProcessed);
+
+        serde_json::to_vec(&metric)
```
Comment (lukesteensen) [inline src/transforms/metric_to_log.rs:66]: Instead of going all the way to bytes, it'd probably be faster to use `to_value`.

**Comment by drunkirishcoder** [comment]
@bruceg @lukesteensen addressed both comments.

**Comment by drunkirishcoder** [comment]
test failure seems to be unrelated to the code in this PR.

**Comment by jamtur01** [comment]
I think the test failures here of:

```
sinks::console::test::encodes_histogram_without_timestamp
sinks::console::test::encodes_set
```

Probably need a quick look before this is merged?

1. sinks::console::test::encodes_histogram_without_timestamp

```
thread 'main' panicked at 'assertion failed: `(left == right)`
  left: `"{\"name\":\"glork\",\"timestamp\":null,\"tags\":null,\"kind\":\"incremental\",\"distribution\":{\"values\":[10.0],\"sample_rates\":[1],\"statistic\":\"histogram\"}}"`,
 right: `"{\"name\":\"glork\",\"kind\":\"incremental\",\"distribution\":{\"values\":[10.0],\"sample_rates\":[1],\"statistic\":\"histogram\"}}"`', src/sinks/console.rs:223:9
stack backtrace:
   0:     0x5640e9110d35 - backtrace::backtrace::libunwind::trace::h14d338b30b3ea0a7
                               at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.46/src/backtrace/libunwind.rs:86
   1:     0x5640e9110d35 - backtrace::backtrace::trace_unsynchronized::h73ea91d74a3fd67f
                               at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.46/src/backtrace/mod.rs:66
   2:     0x5640e9110d35 - std::sys_common::backtrace::_print_fmt::hd42948c952866e12
                               at src/libstd/sys_common/backtrace.rs:78
   3:     0x5640e9110d35 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::ha8f928866ff7571e
                               at src/libstd/sys_common/backtrace.rs:59
   4:     0x5640e914ccbc - core::fmt::write::he0c1e5f7426d2718
                               at src/libcore/fmt/mod.rs:1076
   5:     0x5640e9102ab2 - std::io::Write::write_fmt::hf3afc6cfd57d0033
                               at src/libstd/io/mod.rs:1537
   6:     0x5640e9115bd0 - std::sys_common::backtrace::_print::hfc0110703f3696fd
                               at src/libstd/sys_common/backtrace.rs:62
   7:     0x5640e9115bd0 - std::sys_common::backtrace::print::h3f77c6990ddfaa22
                               at src/libstd/sys_common/backtrace.rs:49
   8:     0x5640e9115bd0 - std::panicking::default_hook::{{closure}}::heae49580a8d62d75
                               at src/libstd/panicking.rs:198
   9:     0x5640e911591c - std::panicking::default_hook::hecc34e3f729e213c
                               at src/libstd/panicking.rs:217
  10:     0x5640e9116323 - std::panicking::rust_panic_with_hook::he82f5d0644692441
                               at src/libstd/panicking.rs:526
  11:     0x5640e9115f1b - rust_begin_unwind
                               at src/libstd/panicking.rs:437
  12:     0x5640e9115e8b - std::panicking::begin_panic_fmt::h905a6d44880d49ef
                               at src/libstd/panicking.rs:391
  13:     0x5640e1b8ca6a - vector::sinks::console::test::encodes_histogram_without_timestamp::h7c2b260cd9c6381f
                               at src/sinks/console.rs:223
  14:     0x5640e1b8c5fa - vector::sinks::console::test::encodes_histogram_without_timestamp::{{closure}}::hea604a49fdbcd5b1
                               at src/sinks/console.rs:211
  15:     0x5640e3eb293e - core::ops::function::FnOnce::call_once::h0f4a70bc901abdb9
                               at /rustc/04488afe34512aa4c33566eb16d8c912a3ae04f9/src/libcore/ops/function.rs:233
  16:     0x5640e501f51c - <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once::h6633cb15d0d76942
                               at /rustc/04488afe34512aa4c33566eb16d8c912a3ae04f9/src/liballoc/boxed.rs:1081
  17:     0x5640e501f51c - <std::panic::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once::h1e20120def172c5c
                               at /rustc/04488afe34512aa4c33566eb16d8c912a3ae04f9/src/libstd/panic.rs:318
  18:     0x5640e501f51c - std::panicking::try::do_call::hcc2ec3bbb75e9316
                               at /rustc/04488afe34512aa4c33566eb16d8c912a3ae04f9/src/libstd/panicking.rs:348
  19:     0x5640e501f51c - std::panicking::try::h94eaebaaa7dd6f41
                               at /rustc/04488afe34512aa4c33566eb16d8c912a3ae04f9/src/libstd/panicking.rs:325
  20:     0x5640e501f51c - std::panic::catch_unwind::h151c07c08497cf8b
                               at /rustc/04488afe34512aa4c33566eb16d8c912a3ae04f9/src/libstd/panic.rs:394
  21:     0x5640e501f51c - test::run_test_in_process::hd082de93b1922c89
                               at src/libtest/lib.rs:541
  22:     0x5640e501f51c - test::run_test::run_test_inner::{{closure}}::h22369c9424e5ab3a
                               at src/libtest/lib.rs:450
  23:     0x5640e501ebdc - test::run_test::run_test_inner::h7171632e89596f84
                               at src/libtest/lib.rs:475
  24:     0x5640e501cdfe - test::run_test::hf31047d4eeefb29d
                               at src/libtest/lib.rs:505
  25:     0x5640e500840e - test::run_tests::h62687325dfd4f954
                               at src/libtest/lib.rs:284
  26:     0x5640e500840e - test::console::run_tests_console::h3261a1dce58028b7
                               at src/libtest/console.rs:280
  27:     0x5640e5018afa - test::test_main::h40987b205cfb5fac
                               at src/libtest/lib.rs:120
  28:     0x5640e501a02d - test::test_main_static::h57077118656a811f
                               at src/libtest/lib.rs:139
  29:     0x5640e24af116 - vector::main::hb427769bafd54f87
  30:     0x5640e3e9dc4b - std::rt::lang_start::{{closure}}::h4eda5a93fae6eec8
                               at /rustc/04488afe34512aa4c33566eb16d8c912a3ae04f9/src/libstd/rt.rs:67
  31:     0x5640e91167c3 - std::rt::lang_start_internal::{{closure}}::h5d3ea623498f5f43
                               at src/libstd/rt.rs:52
  32:     0x5640e91167c3 - std::panicking::try::do_call::hac65e71be769a440
                               at src/libstd/panicking.rs:348
  33:     0x5640e91167c3 - std::panicking::try::hd4706e264bcf6712
                               at src/libstd/panicking.rs:325
  34:     0x5640e91167c3 - std::panic::catch_unwind::h948a0fb4a8b3ee82
                               at src/libstd/panic.rs:394
  35:     0x5640e91167c3 - std::rt::lang_start_internal::h72cc068ed2d0ac53
                               at src/libstd/rt.rs:51
  36:     0x5640e3e9dc27 - std::rt::lang_start::h6279d4b8d096d68b
                               at /rustc/04488afe34512aa4c33566eb16d8c912a3ae04f9/src/libstd/rt.rs:67
  37:     0x5640e24af14a - main
  38:     0x7fe5ce4970b3 - __libc_start_main
  39:     0x5640e1710f4e - _start
  40:                0x0 - <unknown>
thread 'main' panicked at 'assertion failed: `(left == right)`
  left: `"{\"name\":\"users\",\"timestamp\":null,\"tags\":null,\"kind\":\"incremental\",\"set\":{\"values\":[\"bob\"]}}"`,
 right: `"{\"name\":\"users\",\"kind\":\"incremental\",\"set\":{\"values\":[\"bob\"]}}"`', src/sinks/console.rs:204:9
stack backtrace:
   0:     0x5640e9110d35 - backtrace::backtrace::libunwind::trace::h14d338b30b3ea0a7
                               at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.46/src/backtrace/libunwind.rs:86
   1:     0x5640e9110d35 - backtrace::backtrace::trace_unsynchronized::h73ea91d74a3fd67f
                               at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.46/src/backtrace/mod.rs:66
   2:     0x5640e9110d35 - std::sys_common::backtrace::_print_fmt::hd42948c952866e12
                               at src/libstd/sys_common/backtrace.rs:78
   3:     0x5640e9110d35 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::ha8f928866ff7571e
                               at src/libstd/sys_common/backtrace.rs:59
   4:     0x5640e914ccbc - core::fmt::write::he0c1e5f7426d2718
                               at src/libcore/fmt/mod.rs:1076
   5:     0x5640e9102ab2 - std::io::Write::write_fmt::hf3afc6cfd57d0033
                               at src/libstd/io/mod.rs:1537
   6:     0x5640e9115bd0 - std::sys_common::backtrace::_print::hfc0110703f3696fd
                               at src/libstd/sys_common/backtrace.rs:62
   7:     0x5640e9115bd0 - std::sys_common::backtrace::print::h3f77c6990ddfaa22
                               at src/libstd/sys_common/backtrace.rs:49
   8:     0x5640e9115bd0 - std::panicking::default_hook::{{closure}}::heae49580a8d62d75
                               at src/libstd/panicking.rs:198
   9:     0x5640e911591c - std::panicking::default_hook::hecc34e3f729e213c
                               at src/libstd/panicking.rs:217
  10:     0x5640e9116323 - std::panicking::rust_panic_with_hook::he82f5d0644692441
                               at src/libstd/panicking.rs:526
  11:     0x5640e9115f1b - rust_begin_unwind
test sinks::console::test::encodes_histogram_without_timestamp ... FAILED
```

2. sinks::console::test::encodes_set

```
                              at src/libstd/panicking.rs:437
  12:     0x5640e9115e8b - std::panicking::begin_panic_fmt::h905a6d44880d49ef
                               at src/libstd/panicking.rs:391
  13:     0x5640e1b8c557 - vector::sinks::console::test::encodes_set::h9c33ec6d92b4b3c8
                               at src/sinks/console.rs:204
  14:     0x5640e1b8c0da - vector::sinks::console::test::encodes_set::{{closure}}::h9e8d883c0e204ba1
                               at src/sinks/console.rs:194
  15:     0x5640e3ef272e - core::ops::function::FnOnce::call_once::he08702df6c25c0f9
                               at /rustc/04488afe34512aa4c33566eb16d8c912a3ae04f9/src/libcore/ops/function.rs:233
  16:     0x5640e501f51c - <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once::h6633cb15d0d76942
                               at /rustc/04488afe34512aa4c33566eb16d8c912a3ae04f9/src/liballoc/boxed.rs:1081
  17:     0x5640e501f51c - <std::panic::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once::h1e20120def172c5c
                               at /rustc/04488afe34512aa4c33566eb16d8c912a3ae04f9/src/libstd/panic.rs:318
  18:     0x5640e501f51c - std::panicking::try::do_call::hcc2ec3bbb75e9316
                               at /rustc/04488afe34512aa4c33566eb16d8c912a3ae04f9/src/libstd/panicking.rs:348
  19:     0x5640e501f51c - std::panicking::try::h94eaebaaa7dd6f41
                               at /rustc/04488afe34512aa4c33566eb16d8c912a3ae04f9/src/libstd/panicking.rs:325
  20:     0x5640e501f51c - std::panic::catch_unwind::h151c07c08497cf8b
                               at /rustc/04488afe34512aa4c33566eb16d8c912a3ae04f9/src/libstd/panic.rs:394
  21:     0x5640e501f51c - test::run_test_in_process::hd082de93b1922c89
                               at src/libtest/lib.rs:541
  22:     0x5640e501f51c - test::run_test::run_test_inner::{{closure}}::h22369c9424e5ab3a
                               at src/libtest/lib.rs:450
  23:     0x5640e501ebdc - test::run_test::run_test_inner::h7171632e89596f84
                               at src/libtest/lib.rs:475
  24:     0x5640e501cdfe - test::run_test::hf31047d4eeefb29d
                               at src/libtest/lib.rs:505
  25:     0x5640e500840e - test::run_tests::h62687325dfd4f954
                               at src/libtest/lib.rs:284
  26:     0x5640e500840e - test::console::run_tests_console::h3261a1dce58028b7
                               at src/libtest/console.rs:280
  27:     0x5640e5018afa - test::test_main::h40987b205cfb5fac
                               at src/libtest/lib.rs:120
  28:     0x5640e501a02d - test::test_main_static::h57077118656a811f
                               at src/libtest/lib.rs:139
  29:     0x5640e24af116 - vector::main::hb427769bafd54f87
  30:     0x5640e3e9dc4b - std::rt::lang_start::{{closure}}::h4eda5a93fae6eec8
                               at /rustc/04488afe34512aa4c33566eb16d8c912a3ae04f9/src/libstd/rt.rs:67
  31:     0x5640e91167c3 - std::rt::lang_start_internal::{{closure}}::h5d3ea623498f5f43
                               at src/libstd/rt.rs:52
  32:     0x5640e91167c3 - std::panicking::try::do_call::hac65e71be769a440
                               at src/libstd/panicking.rs:348
  33:     0x5640e91167c3 - std::panicking::try::hd4706e264bcf6712
                               at src/libstd/panicking.rs:325
  34:     0x5640e91167c3 - std::panic::catch_unwind::h948a0fb4a8b3ee82
                               at src/libstd/panic.rs:394
  35:     0x5640e91167c3 - std::rt::lang_start_internal::h72cc068ed2d0ac53
                               at src/libstd/rt.rs:51
  36:     0x5640e3e9dc27 - std::rt::lang_start::h6279d4b8d096d68b
                               at /rustc/04488afe34512aa4c33566eb16d8c912a3ae04f9/src/libstd/rt.rs:67
  37:     0x5640e24af14a - main
  38:     0x7fe5ce4970b3 - __libc_start_main
  39:     0x5640e1710f4e - _start
  40:                0x0 - <unknown>
test sinks::console::test::encodes_set ... FAILED
```

**Comment by jszwedko** [comment]
I think #3984 could be relevant here too.

**Comment by drunkirishcoder** [comment]
@jamtur01 thank you for catching that. hope all good now.

**Comment by jamtur01** [comment]
@bruceg If you're still good with this - we'll get it merged.

**Review by bruceg — APPROVED** [review]

**Comment by jamtur01** [comment]
@drunkirishcoder Thank you! Merged! Awesome contribution. 🎆 🍾 💥

**Pull request merged**